docs/internal/DistributedArchitectureGuide.md

```diff
@@ -10,20 +10,43 @@
 
 ### ActionListener
 
-`ActionListener`s are a means off injecting logic into lower layers of the code. They encapsulate a block of code that takes a response
-value -- the `onResponse()` method --, and then that block of code (the `ActionListener`) is passed into a function that will eventually
-execute the code (call `onResponse()`) when a response value is available. `ActionListener`s are used to pass code down to act on a result,
-rather than lower layers returning a result back up to be acted upon by the caller. One of three things can happen to a listener: it can be
-executed in the same thread — e.g. `ActionListener.run()` --; it can be passed off to another thread to be executed; or it can be added to
-a list someplace, to eventually be executed by some service. `ActionListener`s also define `onFailure()` logic, in case an error is
-encountered before a result can be formed.
+`ActionListener` is generally used as a continuation, in the sense of [_continuation-passing
+style_](https://en.wikipedia.org/wiki/Continuation-passing_style) (CPS): an extra argument to a function which defines how to continue the
+computation when the result is available. This is in contrast to _direct style_ which is the more usual style of calling methods that return
+values directly back to the caller so they can continue executing as normal. There's essentially two ways that computation can continue in
+Java (it can return a value or it can throw an exception) which is why `ActionListener` has both an `onResponse()` and an `onFailure()`
+method.
```

Comment:
I think it is an improvement over the existing text. But `CPS` seems kind of exotic! Is there a reason we are not explicitly using the terms synchronous/asynchronous or maybe blocking/non-blocking call, or say `ActionListener` is a callback that can be passed to an async method, or something along those lines? Those seem to be more wide-spread, I feel. Could link to https://en.wikipedia.org/wiki/Asynchronous_method_invocation, for example.

Reply:
(the source might not be the best, but maybe an option if we're looking for something similar to the CSP link)

Reply:
I think we're positively not doing AMI - that means turning a method `T X()` into two methods `AsyncResult BeginX()` and `T EndX(AsyncResult)` where you have to check `AsyncResult#isDone` every now and then and only call `EndX()` once the result is ready. Bleurgh.

I'm curious tho, why do you think CPS is exotic? It's exactly what we're doing. We have to do the CPS translation by hand because nobody has taught Java how to do it for us yet, but mostly when I'm reading code I'm translating it back into direct style in my head.

I did mention blocking a bit here, I'll see if I can get some mention of sync/async in too. They're somewhat orthogonal tho, lots of paths through our code will complete the listener before returning to the caller which means it's at least a little bit synchronous, and async code can involve blocking too (e.g. if it does disk IO).

Reply:
> I think we're positively not doing AMI

Ignore the link. Bad suggestion! :)

> why do you think CPS is exotic? It's exactly what we're doing.

I'm not disagreeing with it, it just doesn't tell me much right away if I don't know the term. I think rather than defining the style, just simplifying that an `ActionListener` is a callback that can be passed around and completed synchronously, or stored and called back later from another thread seems enough, if that is not too much of a simplification. I'd rather see the simplified explanation that might gradually builds up to CPS maybe.  

All I'm saying is that if I can't grok the concept of action listeners from the code, another new term and going through the CPS generic definition to arrive at callback might be overdoing it. Is callback too inaccurate? I feel like that's more widely-known than CPS. 

I agree that  blocking/unblocking might not be the best way of formulating it.

Just an opinion, curious what others think.

Reply:
WDYT about 43dbbed07b8ad2b868c2f8793c897c05a1225aa4?

Reply:
Great! Thanks for reworking it. :)

One small question: Would it fit right after the first paragraph (or at the end of it) to say: if a method receives an action listener, it should not return or throw, but do one of them ONLY via the action listener. Or is that generalization a bit unnecessary? I kind of think of like that most often, but I understand that we don't have a way of enforcing it, and there might be many counter-examples in the code-base.

Reply:
Yep, see e97eb4c3265715a74b8f8ffa4e4e1c4dc3df78bb.